### ISSUE TITLE
Freebayes I/O overload

### ISSUE THREAD

**Issue by MaxUlysse** [SciLifeLab/Sarek#643]
From @alneberg 

We're experiencing a very similar issue. We have yet to confirm that it is the same type of zombie bash processes but we also have jobs running for seemingly forever using a constant 100% cpu of one core. In our case this is a FreeBayes process. 

![screen shot 2018-09-20 at 15 08 20](https://user-images.githubusercontent.com/1250075/45820437-14f70380-bce7-11e8-9a6b-abc4d2a7dc0b.png)

My belief is that the actual job finishes (or potentially crashes) at around 500 min, where the decline in memory occurs. The job works perfectly when I give it a complete node instead:

![screen shot 2018-09-20 at 15 11 12](https://user-images.githubusercontent.com/1250075/45820628-7f0fa880-bce7-11e8-92f8-8a7e9bd93918.png)

Notice how much faster it ran even though it still only used 1 core. This is of course partly due to not needing to use swap, but I don't think that could explain all of it, right? We actually have another instance running still, which is at about 5000 minutes at the moment...

Summary of our error:
 - It seems reproducible (three out of three times, two different users), it is always the same process that gets stuck
 - This process also happens to start using swap since it runs out of memory
 - The process is ran as an individual slurm job.
 - We use the slurm profile

The command ran was:
```
run ~/Sarek/somaticVC.nf \
--sample Preprocessing/Recalibrated/recalibrated.tsv \
-profile slurm \
--project ngi2016004 \
--genome_base /sw/data/uppnex/ToolBox/hg38bundle \
--genome GRCh38 \
--tools FreeBayes,HaplotypeCaller,Manta,Mutect2,strelka \
--strelkaBP \
--containerPath ~/Sarek/containers/ \
--tag 2.1.0
```

I have a hypothesis that I/O overload could be what is triggering these errors: for @marcelm, the overload is caused by a large amount of analysis running at once, and therefore it has not been reproducible on a per-process-basis for him. For us however, the overload is due to the use of swap which is reproducible.

I would be happy to help out with the debugging of this error, but I would need some assistance in how to do so.

Thank you!

**Comment by ewels** [comment]
xref https://github.com/nf-core/rnaseq/issues/93 - also seeing similar things with the methylation pipeline. I'm beginning to think that this could be a nextflow problem in 0.31.1...

**Comment by marcelm** [comment]
This could be thrashing, at least it looks like it: Some process needs more memory than physically available, the kernel starts to swap to disk, but because the process actually *uses* all the memory, it has to do this over and over again. And since we’re replacing fast RAM access with orders of magnitude slower disk access, everything becomes extremely sluggish.

One argument against this theory is that SLURM should have imposed resource limits on the process to prevent it from allocating more memory than physically available.

**Comment by alneberg** [comment]
Aha, that makes perfect sense actually. A detail here is that the config file used is wrong for irma, where we actually have 16G ram per core instead of 8G as it says in the config file. The process is therefore limited to use 8G memory but is not killed because that would require using more than 16? This is related to the use of the --mem flag on Uppmax, which nextflow uses by default but Uppmax recommends not using.

 So basically, this would be fixed by making sure that the --mem flag is always correct.

**Comment by marcelm** [comment]
@alneberg If you SSH into one of the nodes while it’s doing its 100% CPU thing and run `top`, you can inspect the `%Cpu(s)` line at the top and see whether it spends a lot of time in `wa` (waiting for I/O) state. If so, it’s probably thrashing.

**Comment by alneberg** [comment]
Yes I've tried doing so, but ssh:ing into nodes is currently not functioning on Irma. Good to know what to look for when that is working again though. Thank you!

**Comment by marcelm** [comment]
Using a correct `--mem` setting sounds like a good start, but somehow I’d expect the cluster to prevent these situations from arising at all, no matter how you use it. Perhaps the cluster configuration needs adjustment.

**Comment by alneberg** [comment]
Yeah, maybe, but they also recommend not using the `--mem` flag at all so that can be understandable.

**Comment by pontus** [comment]
--mem on UPPMAX is usable if you want to have access to less resources than you would otherwise have (either memory or cores). Use cores to choose how much memory you allocate.

I'd be very surprised if this was thrashing. Rather, it's likely the OOM-killer having run and killed some process, but not enough to kill the job.

**Comment by ewels** [comment]
> _Use cores to choose how much memory you allocate._

We're aware of this, but there's a reason for us to still define the memory: lots of tools have command line flag options to say how much memory they should use. If we don't tell nextflow (and so don't use `--mem`) then we can't insert this information into the script commands. So I figure that the best compromise is to continue using `--mem` but just try to specify the correct amount.

@alneberg / others: how about simply specifying memory based on the number of cores and having a given multiplier for the UPPMAX configs? eg:

```nextflow
params.mem_per_core = 32.GB
process {
    memory: { task.cpus * params.mem_per_core }
    withName: fastqc { cpus: 1 }
    withName: bwa_align { cpus: 8 }
}
```

Would need to check that doing the above does overwrite the other configs (may need to copy the `memory` line for every `withName` process to ensure this), but could be a simpler way to ensure that we always specify the maximum amount of available memory.

**Comment by MaxUlysse** [comment]
Closing due to inactivity and moving to [nf-core/sarek](https://github.com/nf-core/sarek/)
